[csl] use host time for counting CSL synchronization elapsed time

edmont opened the pull request:
Limit periodic radio API calls to `otPlatRadioGetNow` in order to save power in some scenarios.

The impact of the error in the receive window duration calculation is small compared with the power consumption savings obtained with the reduction of radio API calls in serialized platforms.

edmont commented:
I've also improved the conditions for a synchronization even to be considered: tx frame has CSL IE rather than just checking CSL period > 0.

Irving-cl reviewed (changes requested):
The host-radio time sync is used for CSL Transmitter. And the frequency `OPENTHREAD_POSIX_CONFIG_RCP_TIME_SYNC_INTERVAL`. So basically in host+rcp mode, we don't allow the host to get radio time each time it wants to get the time. Instead, it regularly synced with the radio to get the radio time.

Regarding CSL Receiver, I don't think it's supported on host+RCP mode (Considering CSL Reciever is for low power devices, there is no need to run CSL Receiver on posix platform). So in terms of CSL Receiver, host/RCP sync shouldn't happen.

edmont commented:
The changes in this PR only affect CSL Receiver. Notice that `otPlatRadioGetNow` is being called at an interval of each CSL period. This could be costly in terms of time/energy for some platforms, not necessarily host+RCP, but also multi-core chips. Even when there could be some drift between SubMac and Radio which could lead to inaccurate receive window duration, the savings from reducing the number of radio API calls are more beneficial for those platforms.

Do you think a better approach would be to have this change configurable as a build option? This way other platforms wouldn't be affected.

edmont commented:
Thanks @Irving-cl @suveshpratapa, I've added the build option.

> In the case of drift / inaccurate rx window calculation, is it possible to have another configurable option to account for that error?
I've added to the build option documentation the possibility to use CSL uncertainty for this purpose.

Irving-cl commented:
@edmont 

I still have some questions: It seems that `ot::TimerMicro::GetNow` will finally call the same thing as `otPlatRadioGetNow(&GetInstance())` if for non host-RCP mode.

For example, if we use ot-cli-ftd on nrf528xx platform:
* `ot::TimerMicro::GetNow` -> `otPlatAlarmMicroGetNow` -> `nrf5AlarmGetCurrentTime`
* `otPlatRadioGetNow` -> `otPlatTimeGet` -> `nrf5AlarmGetCurrentTime`

So I think the call count are the same. Will the change make a difference in this case?

edmont commented:
@Irving-cl the difference is that for a multi-core architecture the radio API calls end up being serialized, whereas other APIs like the alarm one reside in the host core. Basically, in those architectures there are two instances of `otPlatTimeGet`, one in the host core and one in the radio core.

Irving-cl reviewed (approved):
Thanks for the explanation! @edmont 

LTGM. Could you add the background of multicore and difference in the comment of `OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_LOCAL_TIME_SYNC`.

hemanth-silabs reviewed (approved):
LGTM 👍

jwhui commented on src/core/config/mac.h:
```diff
@@ -425,6 +425,28 @@
+ * short receive window, CSL uncertainty can be increased.
+ *
+ * This setting can be useful for platforms in which is important to reduce the number of
+ * radio API calls, for instance when they are costly. One typicall situation is a multicore
```
```suggestion
 * radio API calls, for instance when they are costly. One typical situation is a multicore
```

abtink commented on src/core/mac/sub_mac.cpp:
```diff
@@ -352,7 +352,11 @@ void SubMac::HandleReceiveDone(RxFrame *aFrame, Error aError)
         if ((mCslPeriod > 0) && aFrame->mInfo.mRxInfo.mAckedWithSecEnhAck)
         {
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_LOCAL_TIME_SYNC
+            mCslLastSync = ot::TimerMicro::GetNow();
```
I think we can remove `ot::` and use `TimerMicor::GetNow()` directly?
(since we are already in `ot` namespace).

abtink commented on src/core/mac/sub_mac.cpp:
```diff
@@ -1232,9 +1240,13 @@ void SubMac::HandleCslTimer(void)
+    uint32_t curTime, elapsed, semiWindow;
+
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_LOCAL_TIME_SYNC
+    curTime = ot::TimerMicro::GetNow();
```
Same here? Can remove `ot::`?

abtink reviewed (approved):
LGTM.  👍 
Couple of smaller suggestions.
